### PR TITLE
experiment default metrics pixels

### PR DESCRIPTION
**Pull request by SabrinaTardio**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/1204186595873227/1208775176602791/f
iOS PR:  in prep
macOS PR:  https://github.com/duckduckgo/macos-browser/pull/3622
What kind of version bump will this require?: N/A

**Optional**:

Tech Design URL: https://app.asana.com/0/1204186595873227/1208682592686299
CC: @bwaresiak 

**Description**: Implements PixelExperimentKit to provide Experiment metrics API to clients

**Steps to test this PR**:
1. Check tests pass

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16
* [ ] macOS 10.15
* [ ] macOS 11
* [ ] macOS 12

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Review by jaceklyp — CHANGES_REQUESTED** [review]

File: Sources/PixelKit/Extensions/Disctionary+PixelKit.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+public extension Dictionary where Key: Comparable {
+    func toString(pairSeparator: String = ":", entrySeparator: String = ",") -> String {
+        return self.sorted(by: { $0.key < $1.key })
```
[the remark this answers is not in the document]
Comment (jaceklyp) [reply Sources/PixelKit/Extensions/Disctionary+PixelKit.swift:21]: I would actually rename it to just DictionaryExtension or something similar

File: Sources/PixelExperimentKit/PixelExperimentKit.swift [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    ///   must occur before the pixel is fired.
+    public static func fireAppRetentionExperimentPixels() {
+        let valueConversionDictionary: [Int: [ClosedRange<Int>]] = [
+            1: [1...1, 2...2, 3...3, 4...4, 5...5, 6...6, 7...7, 5...7],
+            4: [5...7, 8...15],
+            6: [5...7, 8...15],
+            11: [5...7, 8...15],
+            21: [5...7, 8...15],
+            30: [5...7, 8...15]
+        ]
```
[the remark this answers is not in the document]
Comment (SabrinaTardio) [reply Sources/PixelExperimentKit/PixelExperimentKit.swift:138]: So far the only difference is value 0 present in search but not in app retention.
This values were agreed with Data Science when the Android implementation was done.

They might change/ more will be added according to the need

File: Sources/PixelExperimentKit/ExperimentEventTracker.swift [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    /// 2. If `isInWindow` is `true`, it increments the count for the key.
+    /// 3. If the updated count meets or exceeds the specified `threshold`, the stored count is removed, and the method returns `true`.
+    /// 4. If the updated count does not meet the threshold, it updates the count and returns `false`.
+    ///
+    /// - Parameters:
+    ///   - key: The key used to store and retrieve the count.
+    ///   - threshold: The count threshold that triggers a return of `true`.
+    ///   - isInWindow: A flag indicating if the count should be considered (e.g., within a time window).
+    /// - Returns: `true` if the threshold is exceeded and the count is reset, otherwise `false`.
+    func incrementAndCheckThreshold(forKey key: String, threshold: Int, isInWindow: Bool) -> Bool
```
Comment (jaceklyp) [inline Sources/PixelExperimentKit/ExperimentEventTracker.swift:41]: I think we could improve readability of this one by introducing typealias, e.g.
typealias ThresholdCheckResult = Bool
func updateCount(forKey key: String, threshold: Int, isInWindow: Bool) -> ThresholdCheckResult
WDYT?

File: Sources/PixelExperimentKit/ExperimentEventTracker.swift [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+public struct ExperimentEventTracker: ExperimentEventTracking {
+    let store: ExperimentActionPixelStore
```
Comment (jaceklyp) [inline Sources/PixelExperimentKit/ExperimentEventTracker.swift:45]: can it be private?

File: Sources/PixelExperimentKit/PixelExperimentKit.swift [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    /// 3. Tracks actions performed and sends the pixel once the target value is reached (if applicable).
+    public static func fireExperimentPixel(for subfeatureID: SubfeatureID, metric: String, conversionWindowDays: ClosedRange<Int>, value: String) {
+        // Check is active experiment for user
+        guard let featureFlagger = ExperimentConfig.featureFlagger else {
+            assertionFailure("PixelKit is not configured for experiments")
+            return
+        }
+        guard let experimentData = featureFlagger.getAllActiveExperiments()[subfeatureID] else { return }
+
+        fireExperimentPixelForActiveExperiment(subfeatureID, experimentData: experimentData, metric: metric, conversionWindowDays: conversionWindowDays, value: value)
```
Comment (jaceklyp) [inline Sources/PixelExperimentKit/PixelExperimentKit.swift:92]: could use better formatting ;)

File: Sources/PixelExperimentKit/PixelExperimentKit.swift [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+                    experimentData: experimentData,
+                    metric: metric,
+                    conversionWindowDays: range,
+                    value: "\(value)"
+                )
+            }
+        }
+    }
+
+    private static func fireExperimentPixelForActiveExperiment(_ subfeatureID: SubfeatureID, experimentData: ExperimentData, metric: String, conversionWindowDays: ClosedRange<Int>, value: String) {
```
Comment (jaceklyp) [inline Sources/PixelExperimentKit/PixelExperimentKit.swift:172]: let's reformat it a bit

**Review by jaceklyp — CHANGES_REQUESTED** [review]
some small things, otherwise looking good :)

File: Sources/PixelExperimentKit/ExperimentEventTracker.swift (hunk 4 above)
Comment (SabrinaTardio) [reply Sources/PixelExperimentKit/ExperimentEventTracker.swift]: No because we need to inject UserDefaults from the clients

**Review by jaceklyp — APPROVED** [review]

File: Sources/PixelKit/PixelKit.swift [hunk 7]
```diff
@@ -204,7 +205,7 @@ public final class PixelKit {
             case Source.macDMG.rawValue, Source.macStore.rawValue:
                 headers[Header.client] = "macOS"
             default:
-                break
+                headers[Header.client] = "macOS"
```
Comment (jaceklyp) [inline Sources/PixelKit/PixelKit.swift:208]: won't it have some side effect?

**Review by jaceklyp — APPROVED** [review]

File: Sources/PixelExperimentKit/ExperimentEventTracker.swift (hunk 4 above)
Comment (jaceklyp) [reply Sources/PixelExperimentKit/ExperimentEventTracker.swift]: doesn't it happen through initializer?

File: Sources/PixelKit/PixelKit.swift (hunk 7 above)
Comment (SabrinaTardio) [reply Sources/PixelKit/PixelKit.swift:208]: Basically so far is always macOS and we have some NetP etc sources that still want macOS header, so basically all I am doing here is making sure that is it comes from iOS (that for now it is only used for the experiment) it has the iOS/ipadOS header… avoiding changes for anything else that is using it currently.
When we do the full migration we will need to adjust it but we need to consider other Sources as well

**Pull request merged**